tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -4027,44 +4027,102 @@ Status ConvertGather(OpConverterParams* params) {
   return Status::OK();
 }
 
-Status ConvertMatMulHelper(OpConverterParams* params,
-                           TRT_TensorOrWeights tensor_input,
-                           TRT_ShapedWeights weights_raw, bool transpose_weight,
-                           string node_name) {
-  if (!tensor_input.is_tensor()) {
-    return errors::InvalidArgument("Input 0 expects tensor");
+Status ConvertFullyConnectedHelper(OpConverterParams* params, nvinfer1::ITensor* tensor_a,
+                       TRT_ShapedWeights weights_raw, bool transpose_b,
+                       string node_name) {
+  // Reshape input to 3D - this will be a no-op unless using int8 precision.
+  auto input_dim = tensor_a->getDimensions();
+  while (input_dim.nbDims < 3) {
+    input_dim.d[input_dim.nbDims++] = 1;
   }
-  nvinfer1::ITensor* tensor = tensor_input.tensor();
+  TF_RETURN_IF_ERROR(params->converter->PrepareTensorForShape(
+      TRT_TensorOrWeights(tensor_a), input_dim, /*validation_only=*/false,
+      &tensor_a));
 
+  // FC layer will transpose weights, so we need to pre-transpose.
   TRT_ShapedWeights weights(weights_raw.TrtDType());
-  if (transpose_weight) {
-    weights = weights_raw;
-  } else {
+  if (!transpose_b) {
     weights = params->weight_store->GetTempWeights(weights_raw);
     ReorderCKtoKC(weights_raw, &weights);
+  } else {
+    weights = weights_raw;
   }
   TRT_ShapedWeights biases(weights.TrtDType());
-
-  int noutput = weights.shape_.d[0];
-
-  auto input_dim = tensor->getDimensions();
-  while (input_dim.nbDims != 3) {
-    input_dim.d[input_dim.nbDims++] = 1;
-  }
-  TF_RETURN_IF_ERROR(params->converter->PrepareTensorForShape(
-      tensor_input, input_dim, /*validation_only=*/false, &tensor));
-
+  const int noutput = weights.shape_.d[0];
   nvinfer1::IFullyConnectedLayer* layer =
       params->converter->network()->addFullyConnected(
-          *tensor, noutput, weights.GetTrtWeights(), biases.GetTrtWeights());
-  TFTRT_RETURN_ERROR_IF_NULLPTR(layer, node_name);
+          *tensor_a, noutput, weights.GetTrtWeights(), biases.GetTrtWeights());
 
+  TFTRT_RETURN_ERROR_IF_NULLPTR(layer, node_name);
   nvinfer1::ITensor* output_tensor = layer->getOutput(0);
+  params->outputs->push_back(TRT_TensorOrWeights(output_tensor));
+
+  // Reshape output to 1D - this will be a no-op unless using int8 precision.
   auto output_dim = output_tensor->getDimensions();
   output_dim.nbDims = 1;
   TF_RETURN_IF_ERROR(params->converter->PrepareTensorForShape(
       TRT_TensorOrWeights(output_tensor), output_dim, /*validation_only=*/false,
       &output_tensor));
+
+  return Status::OK();
+}
+
+Status ConvertMatMulHelper(OpConverterParams* params,
```

Comment:
Please add tests for the following cases (if they're legal):
- One of A and B is tensor and the other is weights
  - A.GetTrtDims().nbDims in {0, 1, 2, 3}, and B.GetTrtDims().nbDims in {0, 1, 2, 3}
- Both of A and B are tensors and nbDims in {0,1,2,3}

Reply:
Based on the TF docs (https://www.tensorflow.org/api_docs/python/tf/linalg/matmul), I think `nbDims` would always be at least 1. I'll add these test cases in the next commit.

Reply:
I've added a small set of test cases to cover the different possible situations.